File: src/main/java/pw/mihou/nexus/features/command/synchronizer/overwrites/defaults/NexusDefaultSynchronizeMethods.java
package pw.mihou.nexus.features.command.synchronizer.overwrites.defaults;

import org.javacord.api.DiscordApi;
import org.javacord.api.entity.server.Server;
import org.javacord.api.interaction.ApplicationCommand;
import org.javacord.api.interaction.SlashCommand;
import org.javacord.api.interaction.SlashCommandBuilder;
import pw.mihou.nexus.Nexus;
import pw.mihou.nexus.core.exceptions.NexusFailedActionException;
import pw.mihou.nexus.features.command.facade.NexusCommand;
import pw.mihou.nexus.features.command.synchronizer.overwrites.NexusSynchronizeMethods;

import java.util.Set;
import java.util.concurrent.CompletableFuture;

public class NexusDefaultSynchronizeMethods implements NexusSynchronizeMethods {

    @Override
    public CompletableFuture<Set<ApplicationCommand>> bulkOverwriteGlobal(DiscordApi shard, Set<SlashCommandBuilder> slashCommands) {
        return shard.bulkOverwriteGlobalApplicationCommands(slashCommands).thenApply(applicationCommands -> {
            Nexus.getLogger().debug("All global commands have been synchronized. [size={}]", applicationCommands.size());
            return applicationCommands;
        });
    }

    @Override
    public CompletableFuture<Set<ApplicationCommand>> bulkOverwriteServer(DiscordApi shard, Set<SlashCommandBuilder> slashCommands, long serverId) {
        return shard.bulkOverwriteServerApplicationCommands(serverId, slashCommands).thenApply(applicationCommands -> {
            Nexus.getLogger().debug("All commands with relation for server {} have been synchronized.", serverId);
            return applicationCommands;
        });
    }

    @Override
    public CompletableFuture<Void> deleteForServer(DiscordApi shard, NexusCommand command, long serverId) {
        if (shard.getServerById(serverId).isEmpty()) {
            return getServerNotFoundErrorFrom(shard, serverId);
        }

        Server server = shard.getServerById(serverId).orElseThrow();
        return server.getSlashCommands().thenCompose(slashCommands -> {
            SlashCommand slashCommand = slashCommands.stream()
                    .filter($command -> $command.getName().equalsIgnoreCase(command.getName()))
                    .findAny()
                    .orElse(null);

            if (slashCommand == null) {
                return CompletableFuture.completedFuture(null);
            }

            return slashCommand.deleteForServer(serverId)
                    .thenAccept((unused) ->
                            Nexus.getLogger().debug("The command ({}) has been removed from the server ({}).",
                                    command.getName(), serverId)
                    );
        });
    }

    @Override
    public CompletableFuture<ApplicationCommand> updateForServer(DiscordApi shard, NexusCommand command, long serverId) {
        if (shard.getServerById(serverId).isEmpty()) {
            return getServerNotFoundErrorFrom(shard, serverId);
        }

        Server server = shard.getServerById(serverId).orElseThrow();
        return server.getSlashCommands().thenCompose(slashCommands -> {
            SlashCommand slashCommand = slashCommands.stream()
                    .filter($command -> $command.getName().equalsIgnoreCase(command.getName()))
                    .findAny()
                    .orElse(null);

            if (slashCommand == null) {
                return createForServer(shard, command, serverId);
            }

            return command.asSlashCommandUpdater(slashCommand.getId())
                    .updateForServer(shard, serverId)
                    .thenApply(resultingCommand -> {
                        Nexus.getLogger().debug("The command ({}) has been updated for the server ({}).", command.getName(), serverId);
                        return resultingCommand;
                    });
        });
    }

    @Override
    public CompletableFuture<ApplicationCommand> createForServer(DiscordApi shard, NexusCommand command, long serverId) {
        return command.asSlashCommand().createForServer(shard, serverId).thenApply(slashCommand -> {
            Nexus.getLogger().debug("The command ({}) has been created for the server ({}).", command.getName(),
                    serverId);
            return slashCommand;
        });
    }

    /**
     * Creates a simple {@link CompletableFuture} with a failed future that indicates the server cannot be found.
     *
     * @return A failed future indicating the server cannot be found.
     * @param <U> The return type intended to contain.
     */
    private <U> CompletableFuture<U> getServerNotFoundErrorFrom(DiscordApi shard, long serverId) {
        return CompletableFuture.failedFuture(
                new NexusFailedActionException(
                        "An action failed for Nexus Synchronizer. The server (" + serverId + ")" +
                        " cannot be found on the shard calculated (" + shard.getCurrentShard() +"). " +
                        "Is the total shard number value wrong?"
                )
        );
    }
}

File: src/main/java/pw/mihou/nexus/features/command/synchronizer/overwrites/defaults/NexusDefaultSynchronizeMethods.kt
package pw.mihou.nexus.features.command.synchronizer.overwrites.defaults

import org.javacord.api.DiscordApi
import org.javacord.api.interaction.ApplicationCommand
import org.javacord.api.interaction.SlashCommand
import org.javacord.api.interaction.SlashCommandBuilder
import pw.mihou.nexus.Nexus
import pw.mihou.nexus.configuration.modules.debug
import pw.mihou.nexus.core.exceptions.NexusFailedActionException
import pw.mihou.nexus.features.command.facade.NexusCommand
import pw.mihou.nexus.features.command.synchronizer.overwrites.NexusSynchronizeMethods
import java.util.concurrent.CompletableFuture

object NexusDefaultSynchronizeMethods : NexusSynchronizeMethods {

    override fun bulkOverwriteGlobal(shard: DiscordApi, slashCommands: Set<SlashCommandBuilder>) =
        shard
            .bulkOverwriteGlobalApplicationCommands(slashCommands)
            .and { Nexus.configuration.loggingTemplates.GLOBAL_COMMANDS_SYNCHRONIZED(it).debug() }

    override fun bulkOverwriteServer(shard: DiscordApi, slashCommands: Set<SlashCommandBuilder>, serverId: Long) =
        shard
            .bulkOverwriteServerApplicationCommands(serverId, slashCommands)
            .and { Nexus.configuration.loggingTemplates.SERVER_COMMANDS_SYNCHRONIZED(serverId, it).debug() }

    override fun deleteForServer(shard: DiscordApi, command: NexusCommand, serverId: Long): CompletableFuture<Void> {
        if (shard.getServerById(serverId).isEmpty) {
            return getServerNotFoundErrorFrom(shard, serverId)
        }

        val server = shard.getServerById(serverId).orElseThrow()
        return server.slashCommands.thenCompose { slashCommands ->
            val slashCommand = find(command, from = slashCommands) ?: return@thenCompose CompletableFuture.completedFuture(null)
            return@thenCompose slashCommand.delete().thenAccept {
                Nexus.configuration.loggingTemplates.SERVER_COMMAND_DELETED(serverId, slashCommand).debug()
            }
        }
    }

    override fun updateForServer(shard: DiscordApi, command: NexusCommand, serverId: Long): CompletableFuture<ApplicationCommand> {
        if (shard.getServerById(serverId).isEmpty) {
            return getServerNotFoundErrorFrom(shard, serverId)
        }

        val server = shard.getServerById(serverId).orElseThrow()
        return server.slashCommands.thenCompose { slashCommands ->
            val slashCommand = find(command, from = slashCommands) ?: return@thenCompose createForServer(shard, command, serverId)
            return@thenCompose command.asSlashCommandUpdater(slashCommand.id)
                .updateForServer(shard, serverId)
                .and { Nexus.configuration.loggingTemplates.SERVER_COMMAND_UPDATED(serverId, it).debug() }
                .thenApply { it as ApplicationCommand }
        }
    }

    override fun createForServer(shard: DiscordApi, command: NexusCommand, serverId: Long) =
        command
            .asSlashCommand()
            .createForServer(shard, serverId)
            .and { Nexus.configuration.loggingTemplates.SERVER_COMMAND_CREATED(serverId, it).debug() }
            .thenApply { it as ApplicationCommand }

    private fun find(command: NexusCommand, from: Set<SlashCommand>): SlashCommand? =
        from.firstOrNull { `$command` -> `$command`.name.equals(command.name, ignoreCase = true) }

    private fun <Type> CompletableFuture<Type>.and(`do`: (Type) -> Unit): CompletableFuture<Type> =
        this.thenApply { `do`(it); return@thenApply it }

    private fun <U> getServerNotFoundErrorFrom(shard: DiscordApi, serverId: Long): CompletableFuture<U> =
        CompletableFuture.failedFuture(NexusFailedActionException(
            "An action failed for Nexus Synchronizer. The server (" + serverId + ")" +
                    " cannot be found on the shard calculated (" + shard.currentShard + "). " +
                    "Is the total shard number value wrong?"
        ))
}